code/datums/extensions/lockable.dm

```diff
@@ -0,0 +1,218 @@
+/datum/extension/lockable
+	base_type = /datum/extension/lockable
+	expected_type = /obj/item
+	flags = EXTENSION_FLAG_IMMEDIATE
+	// Configuration variables
+	var/max_code_length = 5		// Max length of the password.
+
+	// Stateful variables
+	var/locked 			= 1		// Is it locked?
+	var/code 					// The passcode currently inputed.
+	var/l_code 			= null	// The actual code of the lock
+	var/l_set 			= 0		// Whether or not the lock has been set.
+	var/l_setshort 		= 0		// Whether or not the ability to set the lock is broken.
+	var/l_hacking 		= 0		// Whether or not the lock is being hacked.
+	var/emagged 		= 0		// What or not the lock is emagged.
+	var/is_digital_lock = FALSE	// Whether or not the lock is digital, and its ability to be brute forced.
+	var/open			= 0		// Whether or not the lock panel is open.
+	var/error					// Any errors from user input. Temporary.
+
+/datum/extension/lockable/New(holder, var/is_digital = FALSE)
+	..(holder)
+	is_digital_lock = is_digital
+
+/datum/extension/lockable/nano_host()
+	return holder.nano_host()
+
+/datum/extension/lockable/ui_data(mob/user, ui_key)
+	var/list/data[0]
+	data["emagged"] = emagged
+	data["locked"] = locked
+	data["disabled"] = l_setshort || emagged
+	if(emagged)
+		data["error"] = "LOCKING SYSTEM ERROR - 1701"
+	else if(l_setshort)
+		data["error"] = "ALERT: MEMORY SYSTEM ERROR - 6040 201"
+	else if(error)
+		data["error"] = error
+	if((src.l_set == 0) && (!src.emagged) && (!src.l_setshort))
+		data["status"] = "[max_code_length]-DIGIT PASSCODE NOT SET. ENTER NEW PASSCODE."
+	data["input_code"] = !locked ? "*****" : (code ? code : "N/A")
+	return data
+
+/datum/extension/lockable/Topic(href, href_list)
+	. = ..()
+	if(.)
+		return
+	var/mob/user = usr
+	if(!can_interact(user))
+		return
+
+	// Always clear error.
+	error = null
+
+	var/key_num = text2num(href_list["key"])
+	if(isnum(key_num))
+		code += href_list["key"]
+		return TOPIC_REFRESH
```

Comment:
As far as I know `TOPIC_REFRESH` is for `OnTopic()`, won't do anything here unless you explicitly handle it like player prefs do.